File: android/lib/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/UserInfo.java
package org.elastos.sdk.elephantwallet.contact.internal;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;

public class UserInfo extends HumanInfo implements IdentifyCodeInterface {
    public class UserJson extends HumanJson {
        @SerializedName(JsonKey.IdentifyCode)
        IdentifyCode identifyCode;
    }

    public static void SetCurrDevId(String devId) {
        mCurrDevId = devId;
    }

    public static String GetCurrDevId() {
        return mCurrDevId;
    }

    public String getCurrDevCarrierAddr() {
        for(CarrierInfo info: boundCarrierArray) {
            if(info.devInfo.devId.equals(mCurrDevId) == true) {
                return info.usrAddr;
            }
        }
        return null;
    }

    public String getCurrDevCarrierId() {
        for(CarrierInfo info: boundCarrierArray) {
            if(info.devInfo.devId.equals(mCurrDevId) == true) {
                return info.usrId;
            }
        }
        return null;
    }

    public int fromJson(String info) {
        UserJson json = new Gson().fromJson(info, UserJson.class);

        int ret = super.fromJsonObj(json);
        if(ret < 0) {
            return ret;
        }

        ret = identifyCode.fromJsonObj(json.identifyCode);
        if(ret < 0) {
            return ret;
        }

        return 0;
    }

    public String toJson() {
        UserJson json = new UserJson();
        json.identifyCode = new IdentifyCode();

        int ret = super.toJsonObj(json);
        if(ret < 0) {
            return null;
        }

        ret = this.identifyCode.toJsonObj(json.identifyCode);
        if(ret < 0) {
            return null;
        }

        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String val = gson.toJson(json);
        return val;
    }

    public IdentifyCode identifyCode = new IdentifyCode();

    private static String mCurrDevId;
} // class UserInfo

File: ios/sdk/swift/internal/UserInfo.swift

public class UserInfo: HumanInfo {
  public typealias Kind = IdentifyCode.Kind
  
  public class UserJson: HumanInfo.HumanJson {
    var identifyCode: IdentifyCode? = nil
    
    enum CodingKeys : String, CodingKey {
      case identifyCode
      
      var rawValue: String {
        get {
          switch self {
            case .identifyCode: return JsonKey.IdentifyCode
          }
        }
      }
    }
  }

  public class func setCurrDevId(devId: String) {
    mCurrDevId = devId
  }

  public func getCurrDevCarrierAddr() -> String? {
    if UserInfo.mCurrDevId == nil
    || boundCarrierArray == nil {
      return nil
    }
    
    for info in boundCarrierArray! {
      if info.devInfo.devId == UserInfo.mCurrDevId {
        return info.usrAddr
      }
    }
    return nil
  }

  public func getCurrDevCarrierId() -> String? {
    if UserInfo.mCurrDevId == nil
    || boundCarrierArray == nil {
      return nil
    }

    for info in boundCarrierArray! {
      if info.devInfo.devId == UserInfo.mCurrDevId {
        return info.usrId
      }
    }
    return nil
  }

  public func fromJson(info: String) -> Int {
    let decode = try! JSONDecoder().decode(UserJson.self, from: info.data(using: .utf8)!)
  
    var ret = super.fromJsonObj(info: decode)
    if ret < 0 {
        return ret
    }

    if decode.identifyCode != nil {
      ret = identifyCode.fromJsonObj(json: decode.identifyCode!)
      if ret < 0 {
        return ret
      }
    }

    return 0
  }

  public func toJson() -> String? {
    let json = UserJson()
    json.identifyCode = IdentifyCode()

    var ret = super.toJsonObj(json: json)
    if ret < 0  {
      return nil
    }

    if json.identifyCode != nil {
      ret = self.identifyCode.toJsonObj(json: json.identifyCode!)
      if ret < 0 {
        return nil
      }
    }
    
    let encoder = JSONEncoder()
    encoder.outputFormatting = .prettyPrinted
    let encode = try! encoder.encode(json)
    let val = String(data: encode, encoding: .utf8)!
    
    return val
  }

  public let identifyCode = IdentifyCode()

  private static var mCurrDevId: String?
} // class UserInfo

